src/Common/StringSearcher.h

```diff
@@ -660,18 +667,54 @@ class StringSearcher<true, ASCII> : public StringSearcherBase
         if (needle == needle_end)
             return haystack;
 
-        while (haystack < haystack_end)
-        {
 #ifdef __SSE4_1__
-            if (haystack + n <= haystack_end && pageSafe(haystack))
+        /// Here is the quick path when needle_size is 1. Compare the first and second characters if the needle_size >= 2.
+        if (needle + 1 == needle_end)
+        {
+            while (haystack < haystack_end)
             {
-                /// find first character
-                const auto v_haystack = _mm_loadu_si128(reinterpret_cast<const __m128i *>(haystack));
-                const auto v_against_pattern = _mm_cmpeq_epi8(v_haystack, pattern);
+                if (haystack + n <= haystack_end && pageSafe(haystack))
+                {
+                    const auto v_haystack = _mm_loadu_si128(reinterpret_cast<const __m128i *>(haystack));
+                    const auto v_against_pattern = _mm_cmpeq_epi8(v_haystack, first_pattern);
+                    const auto mask = _mm_movemask_epi8(v_against_pattern);
+                    if (mask == 0)
+                    {
+                        haystack += n;
+                        continue;
+                    }
+
+                    const auto offset = __builtin_ctz(mask);
+                    haystack += offset;
 
-                const auto mask = _mm_movemask_epi8(v_against_pattern);
+                    return haystack;
+                }
+
+                if (haystack == haystack_end)
+                    return haystack_end;
+
+                if (*haystack == first)
+                    return haystack;
+
+                ++haystack;
+            }
 
-                /// first character not present in 16 octets starting at `haystack`
+            return haystack_end;
+        }
+#endif
+
+        while (haystack < haystack_end)
+        {
+#ifdef __SSE4_1__
+            /// find first and second characters
+            if ((haystack + 1 + n) <= haystack_end && pageSafe(haystack))
+            {
+                const auto v_haystack_block_first = _mm_loadu_si128(reinterpret_cast<const __m128i *>(haystack));
+                const auto v_haystack_block_second = _mm_loadu_si128(reinterpret_cast<const __m128i *>(haystack + 1));
+                const auto v_against_pattern_first = _mm_cmpeq_epi8(v_haystack_block_first, first_pattern);
+                const auto v_against_pattern_second = _mm_cmpeq_epi8(v_haystack_block_second, second_pattern);
+                const auto mask = _mm_movemask_epi8(_mm_and_si128(v_against_pattern_first, v_against_pattern_second));
+                /// first and second characters not present in 16 octets starting at `haystack`
```

Comment:
Question: Is above code (l. 706ff.) safe in cases where the haystack contains the first and second needle characters but not consecutively. Example:
- needle = 'ab', 1st needle character: 'a', 2nd needle character: 'b'
- haystack = 'xx a yy b zz'.

Reply:
In this example,

first_needle_character_vec = aaaaaaa
second_needle_character_vec = bbbbbbb
haystack_characters_from_1st = xxayybz
haystack_characters_from_2nd = xayybzz
comparison_result_1st = 0010000
comparison_result_2nd = 0000100
comparison_result_combined = 0000000

Then there will no matched 'ab' in haystack.

Reply:
You are right, thanks.